pkg/daemon/daemon.go

```diff
@@ -303,86 +309,98 @@ func ptp4lCreateCmd(nodeProfile *ptpv1.PtpProfile, confFilePath string) *exec.Cm
 }
 
-// cmdRun runs given ptpProcess and wait for errors
+// cmdRun runs given ptpProcess and restarts on errors
 func cmdRun(p *ptpProcess, stdoutToSocket bool) {
-	glog.Infof("Starting %s...", p.name)
-	glog.Infof("%s cmd: %+v", p.name, p.cmd)
 	var c net.Conn
-	var err error
-	defer func() {
-		if stdoutToSocket && c != nil {
-			if err := c.Close(); err != nil {
-				glog.Errorf("closing connection returned error %s", err)
-			}
+	done := make(chan struct{}) // Done setting up logging.  Go ahead and wait for process
+	for {
+		glog.Infof("Starting %s...", p.name)
+		glog.Infof("%s cmd: %+v", p.name, p.cmd)
+
+		//
+		// don't discard process stderr output
+		//
+		p.cmd.Stderr = os.Stderr
+		cmdReader, err := p.cmd.StdoutPipe()
+		if err != nil {
+			glog.Errorf("cmdRun() error creating StdoutPipe for %s: %v", p.name, err)
+			break
 		}
-		p.exitCh <- true
-	}()
-
-	//
-	// don't discard process stderr output
-	//
-	p.cmd.Stderr = os.Stderr
-	cmdReader, err := p.cmd.StdoutPipe()
-	if err != nil {
-		glog.Errorf("cmdRun() error creating StdoutPipe for %s: %v", p.name, err)
-		return
-	}
-
-	done := make(chan struct{})
-
-	if !stdoutToSocket {
-		scanner := bufio.NewScanner(cmdReader)
-		go func() {
-			for scanner.Scan() {
-				output := scanner.Text()
-				fmt.Printf("%s\n", output)
-				extractMetrics(p.configName, p.name, p.ifaces, output)
-			}
-			done <- struct{}{}
-		}()
-	} else {
-		go func() {
-		connect:
-			select {
-			case <-p.exitCh:
+		if !stdoutToSocket {
+			scanner := bufio.NewScanner(cmdReader)
+			go func() {
+				for scanner.Scan() {
+					output := scanner.Text()
+					fmt.Printf("%s\n", output)
+					extractMetrics(p.configName, p.name, p.ifaces, output)
+				}
 				done <- struct{}{}
-			default:
-				c, err = net.Dial("unix", eventSocket)
-				if err != nil {
-					glog.Errorf("error trying to connect to event socket")
-					time.Sleep(connectionRetryInterval)
-					goto connect
+			}()
+		} else {
+			go func() {
+			connect:
+				select {
+				case <-p.exitCh:
+					done <- struct{}{}
+				default:
+					c, err = net.Dial("unix", eventSocket)
+					if err != nil {
+						glog.Errorf("error trying to connect to event socket")
+						time.Sleep(connectionRetryInterval)
+						goto connect
+					}
 				}
-			}
-			scanner := bufio.NewScanner(cmdReader)
-			for scanner.Scan() {
-				output := scanner.Text()
-				out := fmt.Sprintf("%s\n", output)
-				fmt.Printf("%s", out)
-				_, err := c.Write([]byte(out))
-				if err != nil {
-					glog.Errorf("Write error:", err)
-					goto connect
+				scanner := bufio.NewScanner(cmdReader)
+				for scanner.Scan() {
+					output := scanner.Text()
+					out := fmt.Sprintf("%s\n", output)
+					fmt.Printf("%s", out)
+					_, err := c.Write([]byte(out))
+					if err != nil {
+						glog.Errorf("Write error:", err)
+						goto connect
+					}
 				}
+				done <- struct{}{}
+			}()
+		}
+		p.execMutex.Lock() // Don't restart after termination
```

Comment:
I dont think we need this mutex lock here only need it from L385

Reply:
The case it's here for is if ptp4l dies, and the command is recreated at 392, then it loops back here, and when cmdRun is at this line, cmdStop is called at the same time. I don't want it to send the stop to the process before starting it, and then start it anyways. If that race condition happens, even though it is incredibly unlikely, it could leave behind the process to never (until the node is restarted) get cleaned up.